File: app/cliente.py
"""Cliente HTTP da API de manutenção prescritiva (ADR-002).

A interface não importa os módulos do domínio: fala com a API pelos mesmos endpoints que
o supervisório usaria. A separação é o que sustenta a afirmação de que a API é o contrato
de integração — se a tela acessasse o índice diretamente, o desenho seria uma promessa
sem lastro, e um segundo cliente exigiria reescrever a lógica.
"""

from __future__ import annotations

import json
import os
from dataclasses import dataclass, field
from typing import Iterator
from urllib.parse import unquote

import httpx

API_PADRAO = os.getenv("API_URL", "http://127.0.0.1:8000")

#: Análises e conversas podem levar dezenas de segundos em máquina sem GPU dedicada
#: (ADR-013). O tempo limite acompanha essa realidade em vez de fingir que não existe.
TEMPO_LIMITE = httpx.Timeout(300.0, connect=5.0)


class ApiIndisponivel(RuntimeError):
    """A API não respondeu. A interface orienta o usuário a subir o serviço."""


@dataclass
class ClienteApi:
    base: str = API_PADRAO
    ultimo_roteamento: dict = field(default_factory=dict)
    """Cabeçalhos da última resposta em fluxo: caminho, condição, documento e fontes."""

    def _url(self, caminho: str) -> str:
        return f"{self.base.rstrip('/')}{caminho}"

    def disponivel(self) -> bool:
        try:
            with httpx.Client(timeout=httpx.Timeout(3.0)) as cliente:
                return cliente.get(self._url("/openapi.json")).status_code == 200
        except httpx.HTTPError:
            return False

    def _post(self, caminho: str, **kwargs) -> dict:
        try:
            with httpx.Client(timeout=TEMPO_LIMITE) as cliente:
                resposta = cliente.post(self._url(caminho), **kwargs)
        except httpx.HTTPError as erro:
            raise ApiIndisponivel(str(erro)) from erro

        if resposta.status_code >= 400:
            detalhe = resposta.json().get("detail", resposta.text)
            raise ValueError(detalhe if isinstance(detalhe, str) else str(detalhe))
        return resposta.json()

    def _get(self, caminho: str) -> dict | list:
        try:
            with httpx.Client(timeout=TEMPO_LIMITE) as cliente:
                resposta = cliente.get(self._url(caminho))
        except httpx.HTTPError as erro:
            raise ApiIndisponivel(str(erro)) from erro
        resposta.raise_for_status()
        return resposta.json()

    # -- operações -----------------------------------------------------------------

    def analisar(self, evento: dict, pergunta: str | None = None) -> dict:
        parametros = {"pergunta": pergunta} if pergunta else None
        return self._post("/eventos/analisar", json=evento, params=parametros)

    def conversar(
        self, pergunta: str, condicao: str | None, historico: list[dict] | None = None
    ) -> dict:
        return self._post(
            "/chat",
            json={
                "pergunta": pergunta,
                "condicao": condicao,
                "historico": historico or [],
            },
        )

    def conversar_em_fluxo(
        self, pergunta: str, condicao: str | None, historico: list[dict] | None = None
    ) -> Iterator[str]:
        """Transmite a resposta em partes, para que a espera seja acompanhável.

        O roteamento e as citações chegam nos cabeçalhos, antes do primeiro token, e
        ficam em :attr:`ultimo_roteamento`. Consultá-los assim evita repetir a chamada
        só para obter as fontes — o que dobraria o tempo de geração.
        """
        corpo = {"pergunta": pergunta, "condicao": condicao, "historico": historico or []}
        self.ultimo_roteamento = {}
        try:
            with httpx.Client(timeout=TEMPO_LIMITE) as cliente:
                with cliente.stream("POST", self._url("/chat/fluxo"), json=corpo) as resposta:
                    resposta.raise_for_status()
                    self.ultimo_roteamento = {
                        "caminho": resposta.headers.get("x-caminho", ""),
                        "condicao": resposta.headers.get("x-condicao", ""),
                        "documento": resposta.headers.get("x-documento", ""),
                        "fontes": json.loads(unquote(resposta.headers.get("x-fontes", "[]"))),
                    }
                    for parte in resposta.iter_text():
                        if parte:
                            yield parte
        except httpx.HTTPError as erro:
            raise ApiIndisponivel(str(erro)) from erro

    def cadastrar_documento(self, condicao: str, nome: str, conteudo: bytes) -> dict:
        return self._post(
            "/documentos",
            data={"condicao": condicao},
            files={"arquivo": (nome, conteudo, "application/pdf")},
        )

    def cobertura(self) -> list[dict]:
        return self._get("/documentos/cobertura")  # type: ignore[return-value]

    def estatisticas(self) -> dict:
        return self._get("/estatisticas")  # type: ignore[return-value]

File: web/src/api/cliente.ts
/**
 * Cliente HTTP da API de manutenção prescritiva (ADR-002).
 *
 * A interface não reimplementa regra de domínio: usa os mesmos endpoints que um
 * supervisório ou um CMMS usaria.
 */

import type {
  AnaliseEvento,
  CoberturaDocumental,
  DocumentoRegistrado,
  EstadoSistema,
  EventoSensor,
  PainelHistorico,
  RoteamentoFluxo,
} from "./tipos";

// Não há método de chat síncrono: em estação sem GPU dedicada a geração leva dezenas de
// segundos, e a conversa usa exclusivamente o fluxo.

const BASE = import.meta.env.VITE_API_URL ?? "/api";

/** A API não respondeu, ou respondeu erro em endpoint que deveria existir. */
export class ApiIndisponivel extends Error {}

/** A API recusou a requisição por conteúdo inválido — mensagem exibível ao usuário. */
export class RequisicaoRecusada extends Error {}

/**
 * A API respondeu, mas o serviço de modelos não pôde atender. Distinta de
 * `ApiIndisponivel` porque a ação corretiva é subir o Ollama, não a API.
 */
export class ModeloIndisponivel extends Error {}

/**
 * Traduz uma resposta de erro na exceção correspondente — três situações com três ações
 * corretivas: 415/422 é o que foi enviado e merece a mensagem da API; 503 é o serviço de
 * modelos fora do ar; o resto é indisponibilidade da própria API.
 */
async function lancarErro(resposta: Response, caminho: string): Promise<never> {
  if (resposta.status === 415 || resposta.status === 422) {
    const corpo = await resposta.json().catch(() => null);
    throw new RequisicaoRecusada(detalhar(corpo) ?? "Requisição recusada pela API.");
  }
  if (resposta.status === 503) {
    const corpo = await resposta.json().catch(() => null);
    throw new ModeloIndisponivel(
      detalhar(corpo) ?? "O serviço de modelos não está respondendo.",
    );
  }
  throw new ApiIndisponivel(`${caminho} respondeu ${resposta.status}`);
}

async function requisitar<T>(caminho: string, opcoes?: RequestInit): Promise<T> {
  let resposta: Response;
  try {
    resposta = await fetch(`${BASE}${caminho}`, opcoes);
  } catch (erro) {
    throw new ApiIndisponivel(String(erro));
  }
  if (!resposta.ok) await lancarErro(resposta, caminho);
  return resposta.json() as Promise<T>;
}

/** Extrai a mensagem legível do corpo de erro, seja texto ou lista de campos do Pydantic. */
function detalhar(corpo: unknown): string | null {
  if (!corpo || typeof corpo !== "object") return null;
  const detalhe = (corpo as { detail?: unknown }).detail;
  if (typeof detalhe === "string") return detalhe;
  if (Array.isArray(detalhe)) {
    const mensagem = (detalhe[0] as { msg?: string })?.msg;
    const campos = detalhe
      .map((item) => (item as { loc?: unknown[] }).loc?.slice(-1)[0])
      .filter(Boolean)
      .join(", ");
    if (campos && mensagem) return `${mensagem} (${campos})`;
    if (campos) return `Campos inválidos ou ausentes: ${campos}.`;
  }
  return null;
}

export const api = {
  sistema: (sinal?: AbortSignal) => requisitar<EstadoSistema>("/sistema", { signal: sinal }),

  estatisticas: (sinal?: AbortSignal) =>
    requisitar<PainelHistorico>("/estatisticas", { signal: sinal }),

  cobertura: (sinal?: AbortSignal) =>
    requisitar<CoberturaDocumental[]>("/documentos/cobertura", { signal: sinal }),

  analisar: (evento: EventoSensor, pergunta?: string, sinal?: AbortSignal) => {
    const busca = pergunta ? `?pergunta=${encodeURIComponent(pergunta)}` : "";
    return requisitar<AnaliseEvento>(`/eventos/analisar${busca}`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify(evento),
      signal: sinal,
    });
  },

  cadastrarDocumento: (condicao: string, arquivo: File) => {
    const dados = new FormData();
    dados.append("condicao", condicao);
    dados.append("arquivo", arquivo);
    return requisitar<DocumentoRegistrado>("/documentos", { method: "POST", body: dados });
  },

  /**
   * Conversa com a resposta transmitida em partes.
   *
   * Roteamento e citações chegam nos cabeçalhos, antes do primeiro token, o que evita uma
   * segunda chamada só para obter as fontes. Cabeçalhos HTTP são limitados a latin-1 e as
   * citações têm acentuação, daí a decodificação percentual.
   *
   * O sinal de cancelamento é necessário na prática: a geração leva dezenas de segundos, e
   * sair da tela deixaria a leitura do fluxo escrevendo num componente desmontado.
   */
  conversarEmFluxo: async (
    pergunta: string,
    condicao: string | null,
    historico: TurnoConversa[],
    aoReceber: (parcial: string) => void,
    sinal?: AbortSignal,
  ): Promise<RoteamentoFluxo> => {
    let resposta: Response;
    try {
      resposta = await fetch(`${BASE}/chat/fluxo`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ pergunta, condicao, historico }),
        signal: sinal,
      });
    } catch (erro) {
      throw new ApiIndisponivel(String(erro));
    }

    if (!resposta.ok) await lancarErro(resposta, "/chat/fluxo");
    if (!resposta.body) throw new ApiIndisponivel("/chat/fluxo não devolveu corpo");

    let fontes: string[] = [];
    try {
      fontes = JSON.parse(decodeURIComponent(resposta.headers.get("x-fontes") ?? "[]"));
    } catch {
      // Cabeçalho malformado não invalida o texto: segue sem a lista de citações.
      fontes = [];
    }

    const roteamento: RoteamentoFluxo = {
      caminho: (resposta.headers.get("x-caminho") ?? "") as RoteamentoFluxo["caminho"],
      condicao: resposta.headers.get("x-condicao") ?? "",
      documento: resposta.headers.get("x-documento") ?? "",
      fontes,
    };

    const leitor = resposta.body.getReader();
    const decodificador = new TextDecoder();
    let acumulado = "";
    for (;;) {
      const { done, value } = await leitor.read();
      if (done) break;
      acumulado += decodificador.decode(value, { stream: true });
      aoReceber(acumulado);
    }
    // Esvazia o decodificador: um caractere multibyte partido entre dois blocos só é
    // resolvido nesta chamada final.
    acumulado += decodificador.decode();
    aoReceber(acumulado);

    return roteamento;
  },
};

export interface TurnoConversa {
  papel: "usuario" | "assistente";
  conteudo: string;
}
